parity-util-mem/README.md

```diff
@@ -2,6 +2,10 @@
 
 Collection of memory related utilities.
 
+## WARNING
+
+Defining a global allocator outside of this library is only safe for no-std environments or when `estimate-heapsize` is used.
```

Comment:
I think this should be expanded on. 
If I understand the issue correctly, the risk for UB occurs when `#[global_allocator]` is set in more than one place? But this is not possible I think? I actually tried recently, I wanted to use the OS allocator in a benchmark and had to disable `parity-util-mem` before it worked. I must be missing something, but I think this is the place where it should be explained.
1. What does "defining a global allocator outside this library" mean?
2. Why is it ok for a `no-std` consumer to do it?

Reply:
See https://github.com/paritytech/parity-common/issues/364 and https://github.com/paritytech/polkadot/issues/922.
1. Defining a global allocator not via a feature of parity-util-mem.
2. The problem only occurs when we use `malloc_usable_size` from allocators.rs, which we don't in no-std consumers.

I'm happy to expand the warning, just let me know what would be the best wording here.

Reply:
It seems like we need a whole paragraph here and I'm still too confused to write it. :/

In your "turd" example you have two `parity-util-mem`, one that pulls in an allocator and one that doesn't: why is this UB? I'd have thought that "cargo features are additive" would cause the second copy to be compiled with the same feature set by the first crate?

"Defining a global allocator not via a feature of parity-util-mem." – this is what happened [for the parachain WASM upgrade test](https://github.com/paritytech/polkadot/pull/918/files#diff-0d754f08395fefc53a33b67bd7e02a22R31) right? But how did that even compile? Because the crate itself doesn't pull in `parity-util-mem`? 
But when compiled as a dependency where `parity-util-mem` was also present it would have failed no?

Anyway maybe something like this: "When `parity-util-mem` is used as a dependency with any of the allocation features enabled, it *must* be the sole place where a global allocator is set and it *must* be present in the dependency tree with a single version. 
The only exception to this rule is when used in a `no-std` context or when the `estimate-heapsize` feature is used. Unless heeded you risk UB; see discussion [here](… …)."

Reply:
> In your "turd" example you have two parity-util-mem, one that pulls in an allocator and one that doesn't: why is this UB? I'd have thought that "cargo features are additive" would cause the second copy to be compiled with the same feature set by the first crate?

Because cargo allows to have multiple versions of the same crate. So one (incompatible with the other) version doesn't define a global allocator (and uses the system allocator), so it expect a `malloc_usable_size` from the system allocator, and another version defines jemalloc. You see why it could be a problem? I'm not even sure why it compiles, but it certainly shouldn't (hence the PR).

> "Defining a global allocator not via a feature of parity-util-mem." – this is what happened for the parachain WASM upgrade test right? But how did that even compile? Because the crate itself doesn't pull in parity-util-mem?
But when compiled as a dependency where parity-util-mem was also present it would have failed no?

This particular case is safe because it's only defined for no-std.